src/markdown/generator.ml

```diff
@@ -0,0 +1,327 @@
+          in
+          let f (content, doc, (anchor : Odoc_document.Url.t option)) =
+            let doc = match doc with [] -> noop | doc -> block doc in
+            let content =
+              match content with
+              | `D code -> inline code
+              | `N l -> documented_src l nbsp
+            in
+            let anchor = match anchor with Some a -> a.anchor | None -> "" in
+            break ++ break ++ Anchor anchor ++ break ++ item_prop nbsp
+            ++ content ++ break ++ break ++ str "  " ++ doc ++ break ++ break
+          in
+          let l = list ~sep:noop (List.map f lines) in
+          l ++ continue rest)
+
+and subpage { title = _; header = _; items; url = _ } nbsp =
+  let content = items in
+  let surround body =
+    if content = [] then break else break ++ break ++ body ++ break
+  in
+  surround @@ item nbsp content
+
+and item nbsp (l : Item.t list) : Markup.t =
+  match l with
+  | [] -> noop
+  | i :: rest -> (
+      let continue r = if r = [] then noop else break ++ break ++ item nbsp r in
+      match i with
+      | Text b -> block b ++ continue rest
+      | Heading h -> break ++ heading h ++ break ++ continue rest
+      | Declaration { attr = _; anchor; content; doc } ->
+          let nbsp' = "&nbsp; &nbsp; &nbsp;" in
+          let decl = documented_src content (nbsp ^ nbsp') in
+          let doc = match doc with [] -> noop | doc -> block doc ++ break in
+          let anchor = match anchor with Some x -> x.anchor | None -> "" in
+          Anchor anchor ++ break ++ item_prop nbsp ++ decl ++ break ++ break
+          ++ doc ++ continue rest
+      | Include
+          { attr = _; anchor = _; content = { summary; status; content }; doc }
+        ->
+          let d =
+            if inline_subpage status then item nbsp content
+            else
+              let s = source_code summary in
+              match doc with [] -> s | doc -> s ++ break ++ break ++ block doc
+          in
+          d ++ continue rest)
+
+let on_sub subp =
+  match subp with
+  | `Page p -> if Link.should_inline p.Subpage.content.url then Some 1 else None
+  | `Include incl -> if inline_subpage incl.Include.status then Some 0 else None
+
+let rec calc_subpages (generate_links : bool) { Subpage.content; _ } =
+  [ page generate_links content ]
+
+and subpages generate_links i =
+  Utils.flatmap ~f:(calc_subpages generate_links) @@ Doctree.Subpages.compute i
+
+and page generate_links ({ Page.header; items; url; _ } as p) =
+  let header = Shift.compute ~on_sub header in
+  let items = Shift.compute ~on_sub items in
+  let subpages = subpages generate_links p in
+  Block
+    ([ Inline (Link.for_printing url) ]
+    @ [ item "&nbsp; " header ++ item "&nbsp; " items ]
+    @ subpages)
+
+let rec subpage subp =
+  let p = subp.Subpage.content in
+  if Link.should_inline p.url then [] else [ render p ]
+
+and render (p : Page.t) =
+  let content fmt =
+    Format.fprintf fmt "%a" Markup.pp (page !generate_links p)
```

Comment:
I think we should pass around a record instead of a bare boolean:
- If it's named `args` or `conf`, it'll be easier to see its purpose. This is a common pattern.
- It'll be used by future options.

This record is already defined in src/odoc/markdown.ml, I think it should be defined here and never passed as a global variable or constructed more than once.
This global variable is not necessary at all.

Reply:
same as https://github.com/ocaml/odoc/pull/723#discussion_r712841985